src/d_iwad.c

```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
+static int num_steam_libraries = 0;
+
+static void AddSteamLibraryDir(char *dir)
+{
+    if (num_steam_libraries < MAX_STEAM_LIBRARIES)
+    {
+        steam_libraries[num_steam_libraries] = M_StringReplace(dir, "\\\\",
+                                                               "\\");
+        ++num_steam_libraries;
+    }
+}
+
+//Read whole of libraryfolders.vdf as string
+//silently fail is something goes wrong reading the file
+
+static char *GetSteamLibText(char *library_path)
+{
+    char *library_file_buffer;
+    size_t library_file_size;
+    size_t count;
+    char *probe;
+    FILE *library_file;
+
+    probe = M_FileCaseExists(library_path);
+
+    if (probe == NULL)
+    {
+        return NULL;
+    }
+
+    library_file = fopen(probe, "rb");
+
+    if (library_file == NULL)
+    {
+        return NULL;
+    }
+
+    library_file_size = M_FileLength(library_file);
+    library_file_buffer = malloc(library_file_size + 1);
+    count = fread(library_file_buffer, 1, library_file_size, library_file);
+    fclose(library_file);
+
+    if (library_file_size != count)
+    {
+        return NULL;
+    }
+
+    library_file_buffer[library_file_size] = '\0';
+    return library_file_buffer;
+}
+
+static char *GetNextSteamLibToken(char **lib_file_buffer)
+{
+    char *buffer;
+    char *token;
+    size_t i, buff_size;
+    boolean success;
+
+    token = NULL;               //a token is anything between "'s'
+    success = false;            //we want to avoid an incomplete token
+                                //so we use this instead of NULL check
+    buffer = *lib_file_buffer;
+    buff_size = strlen(buffer);
+
+    for (i=0; i<buff_size; ++i)
+    {
+        if (buffer[i] == '"')
+        {
+            //first " is the start of the token
+            if (token == NULL)
+            {
+                //skip the " itself
+                token = &buffer[i+1];
+            }
+            //second " is the end of the token
+            else
+            {
+                //convert last " to \0 to avoid allocating more strings
+                buffer[i] = '\0';
+                success = true;
+                break;
+            }
+        }
+    }
+
+    if(success)
+    {
+        //move lib_file_buffer ahead of last token
+        *lib_file_buffer = &buffer[i+1];
+        return token;
+    }
+    else
+    {
+        return NULL;
+    }
+}
+
+static boolean IsValidSteamLibKey(char* key)
+{
+    size_t i;
+    for (i=0; i<strlen(key); ++i)
+    {
+        if (!isdigit(key[i]))
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static char *GetSteamNextLibPath(char **lib_file_buffer)
+{
+    boolean key, key_valid;
+    char *token;
+    key = key_valid = true;
+
+    for (;;)
+    {
+        token = GetNextSteamLibToken(lib_file_buffer);
+
+        if (token == NULL)
+        {
+            return NULL;
+        }
+
+        //file always starts with this line before key/value pairs
```

Comment:
So in that case, just put a call before the start of the loop in `CheckSteamLibraries()` that calls `GetNextSteamLibToken()` and checks it's equal to "LibraryFolders". Then this function should become much simpler.

Reply:
I've changed how I parse the file. I treat the whole VDF file as key value pairs, and I don't bother checking that it starts with "LibraryFolders". The code is a lot simpler now.